File: Examples/C/Examples/UdpConnection.c
#include "../Helpers.h"
#include "Connection.h"
#include <stdio.h>
#include "Ximu3.h"

void UdpConnection()
{
    if (YesOrNo("Search for connections?"))
    {
        XIMU3_NetworkAnnouncement* const networkAnnouncement = XIMU3_network_announcement_new();

        const XIMU3_NetworkAnnouncementMessages messages = XIMU3_network_announcement_get_messages_after_short_delay(networkAnnouncement);

        XIMU3_network_announcement_free(networkAnnouncement);

        if (messages.length == 0)
        {
            printf("No UDP connections available\n");
            return;
        }
        printf("Found %s %s\n", messages.array[0].device_name, messages.array[0].serial_number);

        const XIMU3_UdpConnectionInfo connectionInfo = XIMU3_network_announcement_message_to_udp_connection_info(messages.array[0]);

        XIMU3_network_announcement_messages_free(messages);

        XIMU3_Connection* const connection = XIMU3_connection_new_udp(connectionInfo);

        Run(connection, XIMU3_udp_connection_info_to_string(connectionInfo));
    }
    else
    {
        const XIMU3_UdpConnectionInfo connectionInfo = (XIMU3_UdpConnectionInfo) {
            .ip_address = "192.168.1.1",
            .send_port = 9000,
            .receive_port = 8000,
        }; // replace with actual connection info

        XIMU3_Connection* const connection = XIMU3_connection_new_udp(connectionInfo);

        Run(connection, XIMU3_udp_connection_info_to_string(connectionInfo));
    }
}

File: Examples/Rust/src/examples/udp_connection.rs
use super::connection;
use crate::helpers;
use std::net::Ipv4Addr;
use ximu3::connection_info::*;
use ximu3::network_announcement::*;

pub fn run() {
    if helpers::yes_or_no("Search for connections?") {
        let network_announcement = match NetworkAnnouncement::new() {
            Ok(network_announcement) => network_announcement,
            Err(error) => {
                println!("Network announcement failed. {error}.");
                return;
            }
        };

        let messages = network_announcement.get_messages_after_short_delay();

        let Some(message) = messages.first() else {
            println!("No UDP connections available");
            return;
        };

        println!("Found {} {}", message.device_name, message.serial_number);

        let connection_info = &ConnectionInfo::UdpConnectionInfo(message.into());

        connection::run(connection_info);
    } else {
        let connection_info = &ConnectionInfo::UdpConnectionInfo(UdpConnectionInfo {
            ip_address: Ipv4Addr::new(192, 168, 1, 1),
            send_port: 9000,
            receive_port: 8000,
        }); // replace with actual connection info

        connection::run(connection_info);
    }
}
